### PR TITLE
feat: [PagoPA-2459] - expired status

### PR DESCRIPTION
**Pull request by alessio-acitelli**
#### List of Changes

- changes to allow updating a debt position in expired status

#### Motivation and Context

- allows you to bring back a debt position in expired status

#### How Has This Been Tested?

- Local manual and junit test
- [DEV integration test](https://github.com/pagopa/pagopa-debt-position/actions/runs/12348851734)

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: src/main/java/it/gov/pagopa/debtposition/controller/pd/crud/api/impl/DebtPositionController.java [hunk 1]
```diff
@@ -66,7 +66,7 @@ public ResponseEntity<PaymentPositionModel> createDebtPosition(String organizati
         debtPosition.setServiceType(serviceType);
 
         ArrayList<String> segCodes = segregationCodes != null ? new ArrayList<>(Arrays.asList(segregationCodes.split(","))) : null;
-        PaymentPosition createdDebtPos = paymentPositionService.create(debtPosition, organizationFiscalCode, toPublish, segCodes);
+        PaymentPosition createdDebtPos = paymentPositionService.create(debtPosition, organizationFiscalCode, toPublish, segCodes, "create");
```
Comment (cap-ang) [inline src/main/java/it/gov/pagopa/debtposition/controller/pd/crud/api/impl/DebtPositionController.java:69]: This string hard-coded could be a static constant?

File: src/main/java/it/gov/pagopa/debtposition/controller/pd/crud/api/impl/DebtPositionController.java [hunk 2]
```diff
@@ -163,7 +163,7 @@ public ResponseEntity<PaymentPositionModel> updateDebtPosition(String organizati
         }
 
         ArrayList<String> segCodes = segregationCodes != null ? new ArrayList<>(Arrays.asList(segregationCodes.split(","))) : null;
-        PaymentPosition updatedDebtPos = paymentPositionService.update(paymentPositionModel, organizationFiscalCode, toPublish, segCodes);
+        PaymentPosition updatedDebtPos = paymentPositionService.update(paymentPositionModel, organizationFiscalCode, toPublish, segCodes, "update");
```
Comment (cap-ang) [inline src/main/java/it/gov/pagopa/debtposition/controller/pd/crud/api/impl/DebtPositionController.java:166]: This string hard-coded could be a static constant?

File: src/main/java/it/gov/pagopa/debtposition/controller/pd/crud/api/impl/DebtPositionController.java [hunk 3]
```diff
@@ -190,7 +190,7 @@ public ResponseEntity<Void> createMultipleDebtPositions(String organizationFisca
                 .collect(Collectors.toList());
 
         ArrayList<String> segCodes = segregationCodes != null ? new ArrayList<>(Arrays.asList(segregationCodes.split(","))) : null;
-        List<PaymentPosition> createdDebtPosList = paymentPositionService.createMultipleDebtPositions(debtPositions, organizationFiscalCode, toPublish, segCodes);
+        List<PaymentPosition> createdDebtPosList = paymentPositionService.createMultipleDebtPositions(debtPositions, organizationFiscalCode, toPublish, segCodes, "create");
```
Comment (cap-ang) [inline src/main/java/it/gov/pagopa/debtposition/controller/pd/crud/api/impl/DebtPositionController.java:193]: This string hard-coded could be a static constant?

File: src/main/java/it/gov/pagopa/debtposition/controller/pd/crud/api/impl/DebtPositionController.java [hunk 4]
```diff
@@ -211,7 +211,7 @@ public ResponseEntity<Void> updateMultipleDebtPositions(String organizationFisca
                                                       .collect(Collectors.toList());
 
         ArrayList<String> segCodes = segregationCodes != null ? new ArrayList<>(Arrays.asList(segregationCodes.split(","))) : null;
-        List<PaymentPosition> updatedDebtPosList = paymentPositionService.updateMultipleDebtPositions(debtPositions, organizationFiscalCode, toPublish, segCodes);
+        List<PaymentPosition> updatedDebtPosList = paymentPositionService.updateMultipleDebtPositions(debtPositions, organizationFiscalCode, toPublish, segCodes, "update");
```
Comment (cap-ang) [inline src/main/java/it/gov/pagopa/debtposition/controller/pd/crud/api/impl/DebtPositionController.java:214]: This string hard-coded could be a static field?

File: src/main/java/it/gov/pagopa/debtposition/util/DebtPositionValidation.java [hunk 5]
```diff
@@ -97,8 +100,16 @@ private static void checkPaymentPositionContentCongruency(final PaymentPosition
...
-                    null == pp.getValidityDate() && po.getDueDate().compareTo(today) < 0) {
+        	if (
+           		 // Case 1: validity_date is not null and due_date < validity_date
+           	    (pp.getValidityDate() != null && po.getDueDate().compareTo(pp.getValidityDate()) < 0) ||
+           	    
+           	    // Case 2: validity_date is null and due_date < current time
+           	    (pp.getValidityDate() == null && po.getDueDate().compareTo(today) < 0) ||
+           	    
+           	    // Case 3: Action is "update" and due_date < current time
+           	    (!ArrayUtils.isEmpty(action) && "update".equalsIgnoreCase(action[0]) && po.getDueDate().compareTo(today) < 0)
```
Comment (cap-ang) [inline src/main/java/it/gov/pagopa/debtposition/util/DebtPositionValidation.java:111]: This string hard-coded could be a static constant?

**Review by cap-ang — CHANGES_REQUESTED** [review]
Great work.

Just request some changes and a question: there is any impact on bulk API? :thx

File: src/main/java/it/gov/pagopa/debtposition/util/PublishPaymentUtil.java [hunk 6]
```diff
@@ -34,7 +36,8 @@ else if (null == ppToPublish.getValidityDate() && ppToPublish.getMinDueDate().is
             throw new AppException(AppError.DEBT_POSITION_PUBLISH_DUE_DATE_MISMATCH, ppToPublish.getOrganizationFiscalCode(), ppToPublish.getIupd());
         }
         // Regola 2 - se era stata prevista una data di inizio validità e la richiesta di pubblicazione viene fatta successivamente a tale data viene rilanciato un errore
-        else if (ppToPublish.getValidityDate().isBefore(currentDate)) {
+        // (PAGOPA-2459 - controllo applicato solo se non è un'operazione di update)
```
[the remark this answers is not in the document]
Comment (alessio-acitelli) [reply src/main/java/it/gov/pagopa/debtposition/util/PublishPaymentUtil.java:39]: a mix of comments (english and italian) has been made. For consistency, where I found the previous ones in italian, I put the new ones in italian

File: src/main/java/it/gov/pagopa/debtposition/util/PublishPaymentUtil.java [hunk 7]
```diff
@@ -10,17 +10,19 @@
...
+    public void publishProcess(PaymentPosition ppToPublish, String... action) {
         LocalDateTime currentDate = LocalDateTime.now(ZoneOffset.UTC);
         ppToPublish.setPublishDate(currentDate);
         ppToPublish.setStatus(DebtPositionStatus.PUBLISHED);
         ppToPublish.setLastUpdatedDate(currentDate);
-        // Regola 3 e Regola 4 - se non era stata prevista una data di inizio validità e la data di pubblicazione è < della min_due_date => sovrascrivo lo stato direttamente a VALID
-        if (null == ppToPublish.getValidityDate() && ppToPublish.getMinDueDate().isAfter(currentDate)) {
-            ppToPublish.setValidityDate(currentDate);
+        // Regola 3 e Regola 4 - se non era stata prevista una data di inizio validità e la data di pubblicazione è < della min_due_date => sovrascrivo lo stato direttamente a VALID 
+        if (null == ppToPublish.getValidityDate() && ppToPublish.getMinDueDate().isAfter(currentDate)) {
```
[the remark this answers is not in the document]
Comment (alessio-acitelli) [reply src/main/java/it/gov/pagopa/debtposition/util/PublishPaymentUtil.java:24]: I don't understand what you mean

File: src/test/java/it/gov/pagopa/debtposition/controller/DebtPositionControllerTest.java [hunk 8]
```diff
@@ -1385,6 +1402,45 @@ void updateDebtPosition_CreateAndPublished_200() throws Exception {
 
 	}
 	
+	@Test
```
[the remark this answers is not in the document]
Comment (alessio-acitelli) [reply src/test/java/it/gov/pagopa/debtposition/controller/DebtPositionControllerTest.java:1405]: Uhmm yes, you are right. I have to study carefully how to do it though since the status change to EXPIRED of a debt position occurs via batch

File: src/main/java/it/gov/pagopa/debtposition/util/PublishPaymentUtil.java [hunk 9]
```diff
@@ -10,17 +10,19 @@
...
-    public void publishProcess(PaymentPosition ppToPublish) {
+    public void publishProcess(PaymentPosition ppToPublish, String... action) {
         LocalDateTime currentDate = LocalDateTime.now(ZoneOffset.UTC);
         ppToPublish.setPublishDate(currentDate);
         ppToPublish.setStatus(DebtPositionStatus.PUBLISHED);
         ppToPublish.setLastUpdatedDate(currentDate);
-        // Regola 3 e Regola 4 - se non era stata prevista una data di inizio validità e la data di pubblicazione è < della min_due_date => sovrascrivo lo stato direttamente a VALID
-        if (null == ppToPublish.getValidityDate() && ppToPublish.getMinDueDate().isAfter(currentDate)) {
-            ppToPublish.setValidityDate(currentDate);
+        // Regola 3 e Regola 4 - se non era stata prevista una data di inizio validità e la data di pubblicazione è < della min_due_date => sovrascrivo lo stato direttamente a VALID
```
[the remark this answers is not in the document]
Comment (alessio-acitelli) [reply src/main/java/it/gov/pagopa/debtposition/util/PublishPaymentUtil.java:23]: I didn't insert this comment, it was already there

File: src/main/java/it/gov/pagopa/debtposition/util/DebtPositionValidation.java (hunk 5 above)
Comment (alessio-acitelli) [reply src/main/java/it/gov/pagopa/debtposition/util/DebtPositionValidation.java:111]: I don't understand what you mean

Comment (cap-ang) [reply src/main/java/it/gov/pagopa/debtposition/util/DebtPositionValidation.java:111]: only an observation on as-is change request

File: src/test/java/it/gov/pagopa/debtposition/scheduler/ExpiredPositionsSchedulerTest.java [hunk 10]
```diff
@@ -201,5 +204,86 @@ void manualChangeDebtPositionStatusToExpiredAfterDueDate()
...
+	@Test
+	void manualChangeDebtPositionStatusToExpiredAndUpdateAllowed() 
+			throws Exception {
+
+		// creo una posizione debitoria (con 'validity date') valorizzando il campo switchToExpired a true -> Lo stato deve passare ad EXPIRED passata la due_date
+		mvc.perform(post("/organizations/SCHEDULEEXPANDUPD_12345678901/debtpositions")
+				.content(TestUtil.toJson(DebtPositionMock.getMock7())).contentType(MediaType.APPLICATION_JSON))
+		.andExpect(status().isCreated());
+
+		// porto in pubblicata lo stato della posizione debitoria
```
Comment (cap-ang) [inline src/test/java/it/gov/pagopa/debtposition/scheduler/ExpiredPositionsSchedulerTest.java:217]: `PUBLISHED`

File: src/test/java/it/gov/pagopa/debtposition/scheduler/ExpiredPositionsSchedulerTest.java [hunk 11]
```diff
@@ -201,5 +204,86 @@ void manualChangeDebtPositionStatusToExpiredAfterDueDate()
...
+		// verifico che lo stato sia passato ad EXPIRED
+		mvc.perform(get("/organizations/SCHEDULEEXPANDUPD_12345678901/debtpositions/12345678901IUPDMOCK3")
+				.contentType(MediaType.APPLICATION_JSON)).andExpect(status().isOk())
+		.andExpect(content().contentType(MediaType.APPLICATION_JSON))
+		.andExpect(MockMvcResultMatchers.jsonPath("$.status")
+				.value(DebtPositionStatus.EXPIRED.toString()));
+		
+		// aggiorno la posizione debitoria con un body che non contiene la 'validity date' e ne richiedo la pubblicazione
+		PaymentPositionDTO pp7 = DebtPositionMock.getMock7();
+		pp7.setValidityDate(null);
```
Comment (cap-ang) [inline src/test/java/it/gov/pagopa/debtposition/scheduler/ExpiredPositionsSchedulerTest.java:270]: Is another test with validityDate prior to now useful? Could be an input of this method with 2 values `(null, now-10m)`

**Review by cap-ang — APPROVED** [review]
LGTM just one more test required

**Comment by cap-ang** [comment]
LGTM

**Review by pasqualespica — APPROVED** [review]
